source/mocha.ts

```diff
@@ -108,13 +125,16 @@ const run = async () => {
     commandLineOptions,
     ...(pathMatch == null ? {} : { pathMatch }),
   });
+  let reporterScript = path.resolve(scriptDir, "AsMochaReporter.js");
+  if (/^\w:/.test(reporterScript)) {
+    // Because Windows drive letters foul up ESM dynamic imports
+    // with ERR_UNSUPPORTED_ESM_URL_SCHEME errors.
+    reporterScript = pathToFileURL(reporterScript).toString();
+  }
   // Clear out the default Line and Summary reporters.
-  resolvedConfig.reporters = [];
+  resolvedConfig.reporters = [reporterScript];
```

Comment:
See https://github.com/tstyche/tstyche-as-mocha/pull/4#issuecomment-2571538977. And also this example from TSTyche repo: https://github.com/tstyche/tstyche/blob/main/examples/example-plugin.tst.js

Perhaps it is worth to simply normalize into URL string here:

```suggestion
  resolvedConfig.reporters = [pathToFileURL(reporterScript).toString()];
```

Reply:
I'm going to rubber duck this out, because it's a little convoluted.

## Use-case

```gherkin
Given a developer using an IDE
And a project configured with `tstyche`, `mocha`, and `tstyche-as-mocha`
And the project has some unit tests which use TSTyche `expect` and `describe`
When the developer adds a run configuration for those tests with tstyche-as-mocha
And the developer runs the configuration
Then tstyche-as-mocha will run TSTyche
And tstyche-as-mocha will translate TSTyche test results into the expected Mocha format
And the developer should be able to see those results in their IDE
```

## Details

The execution chain looks like:

1. IDE thinks it is running Mocha.  Actually runs tstyche-as-mocha:
   ```shell
   node node_modules/tstyche-as-mocha --reporter /some/path/to/IDE/reporter.js --additional --mocha --options
   ```
2. Node.js runs the `bin` script specified in `tstyche-as-mocha/package.json`, which just points to the compiled `mocha.ts`.
3. The `mocha.js` script translates the Mocha CLI options to something TSTyche understands, specifying the path to the (compiled) `AsMochaReporter.js` as the only TSTyche Reporter, and runs TSTyche.
4. TSTyche runs with the provided config, loading up the AsMochaReporter via the path provided to it.
5. The AsMochaReporter takes the path provided by the IDE, to its Mocha Reporter, and loads it.
6. As TSTyche runs, the AsMochaReporter translates its events to Mocha format, passing them on to the IDE's Mocha Reporter.
7. The user sees the TSTyche results.  Magic!

There are several imports happening, with various formats of specifiers:

- The tstyche-as-mocha CLI loads TSTyche.  This is straightforward, as the import specifier is just the package subpath: `tstyche/tstyche`.
- As part of that config, the as-mocha CLI passes the AsMochaReport's import specifier to TSTyche.
   > Hmm.  I wonder if we could be clever here and make the import specifier something another subpath, like `tstyche-as-mocha/reporter`, and *not* a filesystem path?  It _should_ work, as both packages should be installed.  It would break in scenarios where the developer hasn't actually added a dependency on tstyche-as-mocha ... but do we care?
- TSTyche loads the AsMochaReporter, using whatever it's given.
- The AsMochaReporter loads the IDE's Mocha Reporter, using the filesystem path given to the CLI.

The trick for that last one is that these paths will, in most cases, be provided as absolute filesystem paths by the IDE, and completely out of the user's control.  And in many cases, will be completely obscured from the user, as all they did was click the little green "run this test" button in their IDE.  But, ultimately, the AsMochaReporter will need to correctly translate them into import specifiers.  For example, the one from JetBrains on macOS will end up looking like:

```
--reporter "/Users/ricko/Applications/IntelliJ IDEA Ultimate.app/Contents/plugins/nodeJS/js/mocha-intellij/lib/mochaIntellijReporter.js"
```

So I think that gives me two avenues to work on:

1. Set up a subpath in tstyche-as-mocha so its reporter can be accessed via something like `tstyche-as-mocha/reporter` and _not_ a filesystem path.  This will eliminate all that really gross `scriptDir` stuff if it works.  Yay!
2. Add some more unit tests for the AsMochaReport to handle the various types of import specifiers it should be ready to handle.

Reply:
Thanks for the details. Now I understand the internals better. Using subpath sounds like a good idea indeed.

---

By the way, this made me intrigued:

> `--reporter "/Users/ricko/Applications/IntelliJ IDEA Ultimate.app/Contents/plugins/nodeJS/js/mocha-intellij/lib/mochaIntellijReporter.js"`

I did not expect to see IntelliJ using JavaScript. Very surprising. Does this mean IntelliJ IDEs would be able somehow to spawn `tstyche` and they just need to get back report in certain shape? Not today, but I will dig around.

Reply:
In theory, yes.  But I admit, I don't know how they accept submissions for supporting a new reporter.  Maybe an issue in YouTrack?  For example:

https://youtrack.jetbrains.com/issue/WEB-70517/Test-reporter-for-vitest-should-display-more-information-about-errors

For the most part, it's actually a pretty seamless UX for devs.  You just write some code which looks vaguely like a `describe` block and the IDE gives you a set of 🟩▶️ affordances for it:

https://github.com/user-attachments/assets/17d047f7-f326-41fb-93aa-1726d53b3089

You don't even have to configure anything explicitly.  The IDE figures out that your project has mocha installed and just dynamically sets up a run config for it the first time you use one of the affordances.  Same for Jest, Vitest, etc.

Reply:
What about vending [`mocha-intellij`](https://github.com/JetBrains/mocha-intellij) into this repo? (This is the same code they ship, right?) It did not change for eight years. Looks stable.

Might be I missed something. The benefits of vending would be:

- no need to translate the path to Mocha reporter,
- also no need for https://github.com/tstyche/tstyche/issues/400,
- in the future the code could be reworked so that TSTyche reporter could emit all what is needed without delegating the task to `mocha-intellij`.

And some yers later the resulting TSTyche reporter can be used to build native IntelliJ plugin. At the moment, that is just a dream, but I feel happy looking at this plan. What you think?